### ISSUE TITLE
Is there a way to keep the IdentityServer cookie active if my RP has a sliding expiration?

### ISSUE THREAD

**Issue by parkinsona** [IdentityServer/IdentityServer3#1449]
Hi,

I recently implemented check session to allow for single sign out. As part of my change, I changed the cookie expiration on my RP from sliding to non-sliding. This was done to ensure identity server was kept active. However, in doing this my ajax calls will often return 302 (which I handle and raise as 403). When I used sliding expiration, this wasn't an issue.

The reason I would like to keep idsvr active is because we have multiple RPs that can interact with eachother. If they all have sliding cookies, you can end up in this kind of situation:
- User logs in to RP 1 - sliding cookie created for RP1, idsvr cookie created. Eg they both have 1 hour expiry.
- User works for 2 hours, and needs to go to RP2 - At this point they will need to log in to idsvr because its expired. RP1 has sliding expiration, so its not kept idsvr active.

If we use non-sliding we have the following:
- User logs into RP1 - non-sliding cookie create for RP2 (5 mins), idsvr created (1hr sliding)
- User logs into RP2 after 2 hours, RP2 cookie is created, and no login screen required
- In RP1, a user goes to a page that contains ajax calls with 10 seconds left on their current cookie. After 10 seconds the user performs an ajax request, which fails.

If I use sliding expiration in my RP, Is there a way to keep IdentityServer cookie active? 

Alternatively, if RP is expired, and idsvr is current, when coming from an ajax call is it possible to get the RP cookie recreated via owin middleware and continue loading ? 

I know this has been asked a lot of times before, but I've not seen a situation quite like ours yet.

**Comment by parkinsona** [comment]
Does the idsvr.session cookie expire at all? 
From the behaviour I'm seeing, it doesn't look like it does. I can be expired in the idsvr, but check session is still running happily.

**Comment by brockallen** [comment]
Your app's cookie has nothing to do with IdSvr's cookie, so changing it from sliding to non-sliding (or vice versa) doesn't affect IdSvr.

IdSvr's has a cookie options config setting: https://identityserver.github.io/Documentation/docs/configuration/authenticationOptions.html

**Comment by parkinsona** [comment]
Hey Brock,
I know the 2 cookies are separate and that changing a setting in one doesn't affect the other.
However, if you set your RP to be sliding, it doesn't go back to idsvr, so after a while the idsvr one will time out. So from that point of view, they do have an impact on each other.

Keeping the RP sliding, gets around ajax hassles. But how can I keep the idsvr session active so that the single sign in still works when I access a different RP?

**Comment by brockallen** [comment]
Wait, Ajax hassles? You should not use cookies for Ajax/APIs. That's unrelated to your question, though...

So in the client (RP) -- your client determines how long it wants to trust the user as being authenticated. IdSvr uses a 10 hour fixed cookie by default. I guess you could change the IdSvr cookie to sliding, but that probably won't get what you want.

**Comment by parkinsona** [comment]
I did mention ajax hassles briefly in my first post. But yeah, its a symptom of the problem and not the problem itself ;)

I've got my idSvr to use a 1 hour sliding cookie. I would prefer to have a 10 hour fixed cookie, but the client wants to be signed out of everything after 1 hour of inactivity.
If I set my RP to sliding, then it stops refreshing the idsvr cookie because it never re-authenticates while the user is active. 

Is there an easy way in my RP to hit idsvr for a new token, while keeping the RP with a sliding cookie?

**Comment by brockallen** [comment]
So then use the max_age param in the clients to ensure the user has logged in w/i the last hour.

**Comment by parkinsona** [comment]
ahhhhh this sounds promising....
I just read the open id doco about this. So max_age is dependent on auth time. Is auth time the first time you login, or is it updated each time you go to the authorize endpoint?

**Comment by parkinsona** [comment]
I had a playwith max_age and I don't think its what I'm after. Even if I'm active in the RP and I'm reauthenticating with idsvr it will still force me to login after 1 hour. If max_age was based on the last time I'd hit the auth endpoint, then I think it could be handy.

I'm looking for a way to do this after inactivity. 

Perhaps I'm still looking at this the wrong way. The only real reason I want a sliding cookies in my RP is to get around the ajax issue. If my ajax call finds an expired RP cookie, it fails. I can't remember what code it returns off the top of my head, but I have a handler to raise this as a 401 and I then reload the page which will refresh the token. I read in another issue that someone had success using CORs. From an ajax call Is it even possible to have your RP go to the auth endpoint (with a current idsvr session), return a token and recreate the local cookie in the middle ware?
I've traced it through ,and it redirects to Identity server,but no message is received back in the owin config.

If its absolutely not possible, then I think I might just be stuck with a few minor caveats.

**Comment by brockallen** [comment]
So this is all due to an ajax issue with the access token lifetime being too short? Then extend the access token lifetime -- that's easy.

**Comment by parkinsona** [comment]
Essentially that's the issue. However, if I increase the token lifetime, it only delays the issue from happening.

Eg. User logs in to RP & Idsvr and the user gets a 20 minute token.
The user can use the site fine for 20 minutes and if there's no ajax calls once 20 minutes is gone, it'll request a new token with no interruption to the user.
However, if you've been working away for 19:40 seconds, and you go to a page that has ajax. Ifafter 20 seconds you make an ajax call (like a drop down switch), then it attempts to redirect to identity server, but the ajax gear loses it from that point. If you don't handle the ajax unauthorised event, nothing will happen on the drop down selection, if you do choose to handle it and return a 401/403 then the user gets really confused because it says they're expired and they have to reload their current page to get a new token. Annoying for the user if they were in the middle o a few changes.

Hope this is starting to make more sense.

**Comment by brockallen** [comment]
Sure, so make your access token 1h more than your longest session cookie expiration.

**Comment by parkinsona** [comment]
That would work if I was using an API and was using an access token. My ajax call is calling a partial view, so I'm only using the identity token.

I wonder... would it be possible to use an access token to authenticate a partial view action?

**Comment by brockallen** [comment]
I'm not sure I'd recommend using the identity token for Ajax calls -- that's now how the protocol was designed. Access tokens are meant for Ajax calls.

**Comment by parkinsona** [comment]
I wouldn't be able to store the access token in the cookie, because it would be expired in this scenario.
Would I be better of retrieving an access token as part of the ajax call? Or would I need to use your oidc.js library to add it to the local store?

**Comment by parkinsona** [comment]
I've had a look into some more options for this.
Would it be feasible to create another owin mapping, eg "/ajax" that uses "UseIdentityServerBearerTokenAuthentication"?
In this route I would include any partial views to be returned via ajax.

Or is there a way to use BearerTokenAuthentication on a my existing partial view action ?

**Comment by brockallen** [comment]
Yes, that mapped path sounds reasonable.

**Comment by parkinsona** [comment]
Hi Brock,

Thanks for the info. I discussed this option with my collegues and we are going to hold off making a change like that for the moment. 

One of the worries I had with using a sliding cookie in my RP, was that if you were active in the RP, and didn't hit idsvr for an hour then the check session would boot you out. However this does not appear to be the case. It appears the checksession endpoint only looks for a idsvr.session cookie with the same session state value as the one you pass it. It doesn't take into account idsvr expiry, which works for us.

The only caveat with this approach, is that if after 1 hour of constant use you go to another RP, you will need to login to idsvr. But that's acceptable for our needs. 

Thanks again for your good work on idsvr, it's been great :)

**Comment by Disciplemike** [comment]
I don't mean to raise the dead on this one, but I recently found myself in the same situation.  I've got IDSRV 3 running with 3 clients using it:
1) API Application with many endpoints for data retrieval
2) Public-Facing, Authorized website.
3) Identity Management Application for managing Identity users/claims etc.(internal only)

Now, app 1 uses bearer tokens only, and I have no problem there.
App 2 uses cookie auth for all of it's own resources, but I had to brew up some custome methods to get new access tokens each time I requested data from App 1.  That was a pain for a bit, but I think I've got it down.
App 3 use cookie auth for all of it's own resources too, but also has some mapped paths that uses bearer tokens, because it feed's data to another application.  However, this application also has pages with ajax elements (just to avoid posting).  I think I now see the trouble of using cookies for any authorized ajax calls, even if the calls are to the same site.  If the cookie is expired, the calls will just fail, and the user will not be redirected to a login.
My questions here are two fold:
1) Do I need to do the same basic setup on App 3 that I did on App2 where I use access_tokens and make sure I have a current one before making the ajax call?
2) What are some of the good/accepted ways to handle redirecting to the login whenever a client side action fails an authorization?
